src/main/java/org/broadinstitute/hellbender/tools/exome/IntervalBackedExonCollection.java

```diff
@@ -33,46 +63,114 @@
+    }
+
+    /**
+     * Creates a exon data-base give a sorted list of intervals.
+     *
+     * <p>
+     *     The contig name comparator provided is used to establish the absolute order between
+     *     intervals on different contigs.
+     * </p>
+     *
+     * @param intervals the input interval list. Is assumed to be sorted
+     * @param contigComparator comparator used to establish the absolute order
+     *                         between intervals in different contigs.
+     * @throws IllegalArgumentException if {@code intervals} is {@code null}.
+     */
+    public IntervalBackedExonCollection(final List<SimpleInterval> intervals,
+                                        final Comparator<String> contigComparator) {
         if (intervals == null) {
             throw new IllegalArgumentException("the input intervals cannot be null");
         }
-        checkInputIntervalCoherence(intervals);
-        this.sortedIntervals = Collections.unmodifiableList(new ArrayList<>(intervals));
-        this.intervalsByName = composeIntervalsByName(intervals);
+        if (contigComparator == null) {
+            throw new IllegalArgumentException("the contig comparator cannot be null");
+        }
+        if (intervals.contains(null)) {
+            throw new IllegalArgumentException("the input cannot contain null");
+        }
+
+        intervalOverlapComparator = buildIntervalOverlapComparator(contigComparator);
+        intervalStandardComparator = buildIntervalStandardComparator(contigComparator);
+        sortedIntervals = intervals.stream().sorted(intervalStandardComparator).collect(Collectors.toList());
+        checkForOverlaps(sortedIntervals,intervalOverlapComparator);
+        this.intervalsByName = composeIntervalsByName(sortedIntervals);
     }
 
     /**
-     * Throws a exception if there some interval location incoherence.
+     * Fails with an exception if the intervals collection has overlapping intervals.
+     * @param sortedIntervals
+     * @param overlapComparator
+     */
+    private static void checkForOverlaps(final List<SimpleInterval> sortedIntervals,
+                                  final Comparator<SimpleInterval> overlapComparator) {
+        final OptionalInt failureIndex = IntStream.range(1, sortedIntervals.size())
+                .filter(i -> overlapComparator.compare(sortedIntervals.get(i-1),sortedIntervals.get(i)) == 0)
+                .findFirst();
+
+        if (failureIndex.isPresent()) {
+            final int index = failureIndex.getAsInt();
+            throw new IllegalArgumentException(
+                    String.format("input intervals contain at least two overlapping intervals: %s and %s",
+                            sortedIntervals.get(index-1),sortedIntervals.get(index)));
+        }
+    }
+
+    /**
+     * Given a contig ({@link String}) comparator it will compose interval comparator that will
+     * return 0 when both intervals are overlapping.
      *
```

Comment:
add to the javadoc
 Note: this comparator imposes orderings that are inconsistent with equals."